beacon-chain/rpc/eth/events/events.go

```diff
@@ -88,356 +125,407 @@ func (s *Server) StreamEvents(w http.ResponseWriter, r *http.Request) {
-		if ok {
-			return send(w, flusher, AttesterSlashingTopic, structs.AttesterSlashingFromConsensus(slashing))
-		}
-		// TODO: extend to Electra
-	case operation.ProposerSlashingReceived:
-		if _, ok := requestedTopics[ProposerSlashingTopic]; !ok {
-			return nil
-		}
-		proposerSlashingData, ok := event.Data.(*operation.ProposerSlashingReceivedData)
-		if !ok {
-			return write(w, flusher, topicDataMismatch, event.Data, ProposerSlashingTopic)
-		}
-		return send(w, flusher, ProposerSlashingTopic, structs.ProposerSlashingFromConsensus(proposerSlashingData.ProposerSlashing))
+var (
+	ErrSlowReader = errors.New("client failed to read fast enough to keep outgoing buffer below threadhold")
+	ErrFinished   = errors.New("event received after streamer shut down")
+)
+
+// StreamingResponseWriter defines a type that can be used by the eventStreamer.
+// This must be an http.ResponseWriter that supports flushing and hijacking.
+type StreamingResponseWriter interface {
+	http.ResponseWriter
+	http.Flusher
+}
+
+type lazyReader func() io.Reader
+
+type eventStreamer struct {
+	sync.Mutex
+	ctx      context.Context
+	cancel   func()
+	w        StreamingResponseWriter
+	outbox   chan lazyReader
+	finished chan struct{}
+	kaDur    time.Duration
+}
+
+func NewEventStreamer(ctx context.Context, w http.ResponseWriter, buffSize int, kaDur time.Duration) (*eventStreamer, error) {
+	if kaDur == 0 {
+		kaDur = time.Duration(params.BeaconConfig().SecondsPerSlot) * time.Second
 	}
-	return nil
+	f, ok := w.(StreamingResponseWriter)
+	if !ok {
+		return nil, errors.New("beacon node misconfiguration: http stack may not support required response handling features, like flushing")
+	}
+	ctx, cancel := context.WithCancel(ctx)
+	return &eventStreamer{
+		ctx:    ctx,
+		cancel: cancel,
+		w:      f,
+		outbox: make(chan lazyReader, buffSize),
+		kaDur:  kaDur,
+	}, nil
+}
+
+func (es *eventStreamer) Start() {
+	// Set up SSE response headers
+	es.w.Header().Set("Content-Type", api.EventStreamMediaType)
+	es.w.Header().Set("Connection", api.KeepAlive)
+	es.spawnWriteLoop()
 }
 
-func (s *Server) handleStateEvents(ctx context.Context, w http.ResponseWriter, flusher http.Flusher, requestedTopics map[string]bool, event *feed.Event) error {
-	switch event.Type {
-	case statefeed.NewHead:
-		if _, ok := requestedTopics[HeadTopic]; ok {
-			headData, ok := event.Data.(*ethpb.EventHead)
-			if !ok {
-				return write(w, flusher, topicDataMismatch, event.Data, HeadTopic)
+func (es *eventStreamer) Stop() {
+	es.Cleanup(nil)
+}
+
+func newlineReader() io.Reader {
+	return bytes.NewBufferString(":\n\n")
+}
+
+// We main only care about this keep alive value in a real test setups and using a package var allows test
+// to modify it for convenience.
+var keepAliveInterval = time.Duration(12 * time.Second)
+
+func (es *eventStreamer) spawnWriteLoop() {
+	go func() {
+		var err error
+		kaT := time.NewTimer(es.kaDur)
+		defer func() {
```

Comment:
maybe we can merge defers here

Reply:
Could you please add a reason for your suggestion?

The code would be functionally the same since defers execute sequentially (LIFO), the reason I wrote it this way is that I like to immediately follow the setup of a thing in need of cleanup with the defer to do so, ie: make a timer, set up the timer's cleanup.